app/scripts/modules/core/src/cloudProvider/versionedCloudProvider.service.ts

```diff
@@ -24,6 +26,50 @@ export class VersionedCloudProviderService {
       : this.cloudProviderRegistry.getValue(cloudProvider, key);
   }
 
+  public getInstanceProviderVersion(cloudProvider: string, instanceId: string, app: Application): IPromise<string> {
+    const providerVersions = this.accounts.reduce(
+      (versions, account) => {
+        if (account.cloudProvider === cloudProvider && !!account.providerVersion) {
+          versions.add(account.providerVersion);
+        }
+        return versions;
+      },
+      new Set<string>(),
+    );
+
+    if (providerVersions.size === 0) {
+      // Rely on the cloudProviderRegistry to return the default provider implementation.
+      return this.$q.resolve(null);
+    } else if (providerVersions.size === 1) {
+      return this.$q.resolve(Array.from(providerVersions)[0]);
+    }
+
+    return app.ready().then(() => {
```

Comment:
Standalone instances are routed through a separate state handler (which is passed the instance's account, so it will be easy to resolve the provider version): https://github.com/spinnaker/deck/blob/master/app/scripts/modules/core/src/instance/instance.states.ts#L73-L83

I was planning on handling that case in another PR.